File: magi/utils/memory.py
"""
memory.py - Utilities for managing and persisting agent memory
"""

import os
import json
from typing import Dict, Any, List

# Global memory storage to persist between commands
memory: Dict[str, List[Any]] = {
    "input_items": [],  # Store all previous inputs
    "outputs": []       # Store all previous outputs
}

def save_memory() -> None:
    """Save memory to a persistent file."""
    try:
        # Try multiple locations in order of preference
        # 1. Shared volume (might be read-only)
        # 2. Home directory (should be writable)
        # 3. /tmp directory (always writable)
        memory_locations = [
            "/claude_shared",
            os.path.expanduser("~"),
            "/tmp"
        ]

        for memory_dir in memory_locations:
            memory_file = f"{memory_dir}/magi_memory.json"

            # Skip this location if it doesn't exist or isn't writable
            if not os.path.exists(memory_dir) or not os.access(memory_dir, os.W_OK):
                continue

            try:
                # Now save the memory file
                with open(memory_file, "w") as f:
                    json.dump(memory, f)

                # Set permissions on the file
                os.chmod(memory_file, 0o666)

                # Successfully saved, no need to try other locations
                return
            except Exception as inner_e:
                # Try the next location
                continue

        # If we get here, we couldn't save to any location
        print(f"Warning: Could not save memory to any available location")
    except Exception as e:
        print(f"Error saving memory: {str(e)}")
        # Continue execution even if saving fails

def load_memory() -> None:
    """Load memory from persistent storage."""
    global memory
    try:
        # Try the same locations as save_memory in the same order
        memory_locations = [
            "/claude_shared",
            os.path.expanduser("~"),
            "/tmp"
        ]

        for memory_dir in memory_locations:
            memory_file = f"{memory_dir}/magi_memory.json"

            # Skip if file doesn't exist
            if not os.path.exists(memory_file):
                continue

            try:
                with open(memory_file, "r") as f:
                    loaded_memory = json.load(f)
                    memory.update(loaded_memory)
                    return
            except Exception:
                # Try the next location
                continue

        # If we get here, we couldn't load from any location
        print("No memory file found (this is normal for first run)")
    except Exception as e:
        print(f"Error loading memory (this is normal for first run): {str(e)}")

def add_input(input_content: str) -> None:
    """Add an input to memory."""
    memory["input_items"].append({"content": input_content, "type": "text"})

def add_output(output_content: str) -> None:
    """Add an output to memory."""
    memory["outputs"].append(output_content)
    # Save memory after each update
    save_memory()

def get_memory() -> Dict[str, List[Any]]:
    """Get the current memory."""
    return memory

File: magi/src/utils/memory.ts
/**
 * Memory management for the MAGI system.
 *
 * This module provides functions to store and retrieve conversation history
 * across sessions.
 */

import fs from 'fs';
import path from 'path';
import os from 'os';

// Directory to store memory files
const MEMORY_DIR = path.join(os.homedir(), '.magi', 'memory');

// Maximum number of history items to keep
const MAX_HISTORY_ITEMS = 10;

// Memory structure
interface Memory {
    inputs: string[];
    outputs: string[];
    lastAccessed: number;
}

// Global memory cache
let memoryCache: Memory = {
    inputs: [],
    outputs: [],
    lastAccessed: Date.now(),
};

/**
 * Ensure the memory directory exists
 */
function ensureMemoryDir(): void {
    if (!fs.existsSync(MEMORY_DIR)) {
        fs.mkdirSync(MEMORY_DIR, { recursive: true });
    }
}

/**
 * Get the path to the memory file
 */
function getMemoryFilePath(): string {
    ensureMemoryDir();
    return path.join(MEMORY_DIR, 'memory.json');
}

/**
 * Load conversation memory from disk
 */
export function loadMemory(): Memory {
    try {
        const memoryPath = getMemoryFilePath();

        if (fs.existsSync(memoryPath)) {
            const data = fs.readFileSync(memoryPath, 'utf-8');
            memoryCache = JSON.parse(data);
            console.log(
                `Loaded memory with ${memoryCache.inputs.length} inputs and ${memoryCache.outputs.length} outputs`
            );
        } else {
            console.log('No existing memory found, starting fresh');
            memoryCache = { inputs: [], outputs: [], lastAccessed: Date.now() };
            saveMemory();
        }
    } catch (error) {
        console.error(`Error loading memory: ${error}`);
        memoryCache = { inputs: [], outputs: [], lastAccessed: Date.now() };
    }

    return memoryCache;
}

/**
 * Save conversation memory to disk
 */
export function saveMemory(): void {
    try {
        ensureMemoryDir();
        const memoryPath = getMemoryFilePath();

        // Update last accessed time
        memoryCache.lastAccessed = Date.now();

        // Write to disk
        fs.writeFileSync(
            memoryPath,
            JSON.stringify(memoryCache, null, 2),
            'utf-8'
        );

        console.log('Memory saved successfully');
    } catch (error) {
        console.error(`Error saving memory: ${error}`);
    }
}

/**
 * Add user input to memory
 */
export function addInput(input: string): void {
    memoryCache.inputs.push(input);

    // Trim to maximum size
    if (memoryCache.inputs.length > MAX_HISTORY_ITEMS) {
        memoryCache.inputs = memoryCache.inputs.slice(-MAX_HISTORY_ITEMS);
    }

    // Save updates
    saveMemory();
}

/**
 * Add agent output to memory
 */
export function addOutput(output: string): void {
    memoryCache.outputs.push(output);

    // Trim to maximum size
    if (memoryCache.outputs.length > MAX_HISTORY_ITEMS) {
        memoryCache.outputs = memoryCache.outputs.slice(-MAX_HISTORY_ITEMS);
    }

    // Save updates
    saveMemory();
}

/**
 * Get complete conversation history
 */
export function getConversationHistory(): Array<{
    role: string;
    content: string;
}> {
    const history: Array<{ role: string; content: string }> = [];

    // Interleave inputs and outputs
    const maxItems = Math.max(
        memoryCache.inputs.length,
        memoryCache.outputs.length
    );

    for (let i = 0; i < maxItems; i++) {
        // Add user input if available
        if (i < memoryCache.inputs.length) {
            history.push({
                role: 'user',
                content: memoryCache.inputs[i],
            });
        }

        // Add system output if available
        if (i < memoryCache.outputs.length) {
            history.push({
                role: 'assistant',
                content: memoryCache.outputs[i],
            });
        }
    }

    return history;
}

/**
 * Clear all memory
 */
export function clearMemory(): void {
    memoryCache = { inputs: [], outputs: [], lastAccessed: Date.now() };
    saveMemory();
    console.log('Memory cleared successfully');
}
